src/computability/reduce.lean

```diff
@@ -0,0 +1,236 @@
+∃ f, computable f ∧ ∀ a, p a ↔ q (f a)
+
+infix ` ≤₀ `:1000 := many_one_reducible
+
+def one_one_reducible {α} [primcodable α] (p q : α → Prop) := 
+∃ f, computable f ∧ injective f ∧ ∀ a, p a ↔ q (f a)
+
+infix ` ≤₁ `:1000 := one_one_reducible
+
+theorem reflexive_many_one_reducible {α} [primcodable α] : reflexive (@many_one_reducible α _) := 
+λ p, ⟨id, computable.id, by simp⟩
+
+theorem transitive_many_one_reducible {α} [primcodable α] : transitive (@many_one_reducible α _) := 
+λ p q r ⟨f, c₁, h₁⟩ ⟨g, c₂, h₂⟩, 
+⟨g ∘ f, computable.comp c₂ c₁, λ a, 
+  ⟨λ h, by { rwa [←h₂, ←h₁] }, 
+   λ h, by { rwa [h₁, h₂] }⟩⟩
+
+theorem reflexive_one_one_reducible {α} [primcodable α] : reflexive (@one_one_reducible α _) := 
+λ p, ⟨id, computable.id, injective_id, by simp⟩
+
+theorem transitive_one_one_reducible {α} [primcodable α] : transitive (@one_one_reducible α _) := 
+λ p q r ⟨f, c₁, i₁, h₁⟩ ⟨g, c₂, i₂, h₂⟩, 
+⟨g ∘ f, computable.comp c₂ c₁, injective_comp i₂ i₁, λ a,
+  ⟨λ h, by { rwa [←h₂, ←h₁] }, 
+   λ h, by { rwa [h₁, h₂] }⟩⟩
+
+namespace computable_pred
+variables {α : Type*} {σ : Type*}
+variables [primcodable α] [primcodable σ]
+open computable
+
+theorem computable_of_many_one_reducible
+  {p q : α → Prop} 
+  (h₁ : p ≤₀ q) (h₂ : computable_pred q) : computable_pred p := 
+  begin
+    rcases h₁ with ⟨f, c, hf⟩,
+    rcases computable_iff.1 h₂ with ⟨g, hg, rfl⟩, split, 
+    { apply computable.of_eq (cond (comp hg c) (const tt) (const ff)),
+      intro, cases heq : g (f n), repeat { simp [hf, heq] },
+      { intro, cases heq : g (f a), 
+        { left, simp [hf, heq] },
+        { right, simp [hf, heq] } } }
+  end
+
+theorem computable_of_one_one_reducible
+  {p q : α → Prop} 
+  (h₁ : p ≤₁ q) (h₂ : computable_pred q) : computable_pred p := 
+  by { rcases h₁ with ⟨f, c, hinj, hf⟩, apply computable_of_many_one_reducible ⟨f, c, hf⟩ h₂ }
+
+end computable_pred
+
+def many_one_equiv {α} [primcodable α] (p q : α → Prop) := p ≤₀ q ∧ q ≤₀ p
+
+def one_one_equiv {α} [primcodable α] (p q : α → Prop) := p ≤₁ q ∧ q ≤₁ p
+
+theorem equivalence_of_many_one_equiv {α} [primcodable α] : equivalence (@many_one_equiv α _) :=
+⟨λ x,⟨reflexive_many_one_reducible _, reflexive_many_one_reducible _⟩,  
+ λ x y h, ⟨h.2, h.1⟩, 
+ λ x y z h₁ h₂, ⟨transitive_many_one_reducible h₁.1 h₂.1, transitive_many_one_reducible h₂.2 h₁.2⟩⟩
+
+theorem equivalence_of_one_one_equiv {α} [primcodable α] : equivalence (@one_one_equiv α _) :=
+⟨λ x,⟨reflexive_one_one_reducible _, reflexive_one_one_reducible _⟩,  
+ λ x y h, ⟨h.2, h.1⟩, 
+ λ x y z h₁ h₂, ⟨transitive_one_one_reducible h₁.1 h₂.1, transitive_one_one_reducible h₂.2 h₁.2⟩⟩
+
+instance many_one_equiv_setoid {α} [primcodable α] : setoid (set α) := 
+⟨many_one_equiv, @equivalence_of_many_one_equiv α _⟩
+
+def many_one_degree {α} [primcodable α] := quotient (@many_one_equiv_setoid α _)
+
+def disjoin (p q : set ℕ) : set ℕ  := (λ x, 2 * x) '' p ∪ (λ x, 2 * x + 1) '' q
```

Comment:
I would like to see this generalized over other types. Rather than coding everything over nats, this would look much better if the domain is simply a sum type, and the primcodable instance will do all the odd/even interleaving work.

Reply:
I would also like to make things as general as possible, but just like Rice's theorem, the classical results are almost always about nats. I feel that we can go with nats for now, and if the generalized version turns out to be useful for other theories, then we do the refactoring.

Reply:
I actually am claiming that the work you are doing now can be done easier with sum types, even setting aside any possible future benefits of generalization. `sum.inl` is primrec, which replaces your theorems about 2*x and 2*x+1.

Also, `many_one_equiv` and friends can be stated with different types on the left and right, although of course the setoid instance can't.

Reply:
I agree that sum type naturally interleaves things, but then we end up having something like `set (ℕ ⊕ ℕ)` when trying to instantiate them with nats. That looks a bit weird to me when I just want to reason about predicates on natural numbers. We could prove some translation theorems, though.